Replace `return; yield` with `yield from []`

ruudk opened the pull request:
This has the same effect, but looks less hacky and makes PHPStan happy.
https://phpstan.org/r/df7fcc88-1df8-428e-b675-5dc6965c34d6

Previously this was needed to work properly with output capturing.

/cc @fabpot @nicolas-grekas 

https://3v4l.org/KXste

derrabus commented:
I think, the goal is to have a generator function which is not executed until we iterate over the result. This does not happen if you remove all `yield` instructions and return an empty array. But we could do `yield from [];` instead.

ruudk commented:
> This does not happen if you remove a `yield` instructions and return an empty array

Are you sure? https://3v4l.org/GrX71

ruudk commented:
> we could do `yield from [];` instead

I didn't think of this. I like it as well, as this will make sure the return type is always a Generator.

ruudk commented:
Currently, PHPStan does not complain about this 👏 
https://phpstan.org/r/df7fcc88-1df8-428e-b675-5dc6965c34d6

ruudk commented:
I rebased the PR and targeted v3 and applied the `yield from []` trick suggested by @derrabus.

nicolas-grekas commented on src/Node/CaptureNode.php:
```diff
@@ -36,10 +36,10 @@ public function compile(Compiler $compiler): void
         $compiler
             ->raw($useYield ? "implode('', iterator_to_array(" : '\\Twig\\Extension\\CoreExtension::captureOutput(')
-            ->raw("(function () use (&\$context, \$macros, \$blocks) {\n")
+            ->raw("(function () use (&\$context, \$macros, \$blocks): iterable {\n")
```
this is not needed, generated code don't need extra runtime checks, let's remove them all from this PR

ruudk replied:
Done!

nicolas-grekas reviewed (commented):
looks nicer indeed

ruudk committed:
Replace `return; yield` with `yield from []`

This has the same effect, but looks less hacky and makes PHPStan happy.
https://phpstan.org/r/df7fcc88-1df8-428e-b675-5dc6965c34d6

Previously this was needed to work properly with output capturing.

derrabus commented:
> > This does not happen if you remove a `yield` instructions and return an empty array
> 
> Are you sure? https://3v4l.org/GrX71

This reproducer misses the point because the function call and the start of the iteration are on the same line.

see https://3v4l.org/78LhS

ruudk commented:
@nicolas-grekas You approved and I requested review at roughly the same time 😅

fabpot commented:
Thank you @ruudk.
